lib/buildkite/builder/extensions/sub_pipelines.rb

```diff
@@ -16,26 +16,30 @@ def self.to_sym
             name.split('::').last.downcase.to_sym
           end
 
-          def initialize(name, steps, &block)
+          def initialize(name, context, &block)
+            @context = context
             @name = name
             @data = Data.new
             @data.steps = StepCollection.new(
-              steps.templates,
-              steps.plugins
+              context.data.steps.templates,
+              context.data.steps.plugins
             )
             @data.notify = []
             @data.env = {}
 
-            @dsl = Dsl.new(self)
-            @dsl.extend(Extensions::Steps)
-            @dsl.extend(Extensions::Notify)
-            @dsl.extend(Extensions::Env)
+            # Use `clone` to copy over dsl's extended extensions
+            @dsl = context.dsl.clone
+            # Override dsl context to current pipeline
+            @dsl.instance_variable_set(:@context, self)
+
             instance_eval(&block) if block_given?
             self
           end
 
           def to_h
             attributes = super
+            # Merge envs from main pipeline
+            data.env.merge!(@context.data.env)
```

Comment:
The env vars set in this subpipeline should take priority. The env vars from the main pipeline is the fallback/default.